Akolesov disasm options

anthony-kolesov opened the pull request:
Proposition of changes that will allow GDB to select current CPU in disassembler.

1. First patch adds support for `cpu=...` disassembler option that allows to override CPU from mach/private ELF header.
2. Second patch removes duplicate ARC600 entry
3. Third part provides custom bfd_arch_compatible function to satisfy GDB expectations of what it should return for ARC variants.

claziss commented on bfd/cpu-arc.c:
```diff
@@ -53,3 +56,54 @@ static const bfd_arch_info_type arch_info_struct[] =
+  /* ARC600.  */
+  if (a->mach == bfd_mach_arc_arc600 && b->mach == bfd_mach_arc_arc600)
+    return arc600;
+
```
Hi,
we need to handle also the bfd_mach_arc_arc601, and come up with something for the old a4 and a5 configurations (i.e., bfd_mach_arc_a4, and bfd_mach_arc_a5). Otherwise it is alright.

anthony-kolesov replied:
I don't see any `bfd_arch_info_type` definitions for A4 and A5 in `arch_info_struct`, where those can come from as arguments to this function?

claziss replied:
bfd/archures.c

claziss commented on opcodes/arc-dis.c:
```diff
@@ -117,6 +117,11 @@ typedef struct skipclass
+/* ISA mask value enforced via disassembler info options.  ARC_OPCODE_NONE
+   value means that no CPU is enforced.  */
+
+static unsigned enforced_isa_mask = ARC_OPCODE_NONE;
```
we can configure binutils project with a specific CPU (--with-cpu=<name>) is not better to have the configured CPU here?

anthony-kolesov replied:
What would be the use case for this? If option is not specified at all, then disassembler revert to current behavior - checking ELF headers. If instead it will use CPU value that was specified with --with-cpu, then that will change how `objdump -d` works - it will always print disasm for default CPU, ignoring the value in ELF file.

claziss replied:
makes sense

claziss commented on opcodes/arc-dis.c:
```diff
@@ -779,6 +784,35 @@ parse_option (char *option)
+
+static unsigned
+parse_cpu_option (char *option)
+{
```
It will be nice if we can accept the same CPU names as gas. This will keep the consistency among the tools. Maybe in include/elf/ folder to have something like arc-cpu.def which will be shared between gas and disassm?

claziss commented on opcodes/arc-dis.c:
```diff
@@ -1279,6 +1333,16 @@ The following ARC specific disassembler options are supported for use \n\
 with -M switch (multiple options should be separated by commas):\n"));
 
   fprintf (stream, _("\
+  cpu={ARC600,ARC601,A6} Enforce ARC600 ISA.\n"));
```
auto printing based on arc-cpus.def file.

anthony-kolesov commented:
Changes in v2:
- Added support for ARC601, A4, and A5 in `arc_compatible`.
- Added a commit that adds `include/elf/arc-cpu.def`.
- Added its usage to disassembler options.

anthony-kolesov commented:
Pushed another change: argument to `parse_cpu_option` is now `const char` instead of `char`, to avoid build errors on upstream master where, argument to `parse_option` is const.

anthony-kolesov commented:
Corrected typos.

anthony-kolesov commented:
Added an another fix: in arc_compatible() check `if (a == b)`.

claziss commented:
It looks alright. Please make sure there are no regression.

anthony-kolesov commented:
There is an issue with building any uClibc toolchain, where libgcc.so link fails, because for some reason "output" file has architecture ARC600, even though option -mcpu=hs38 is passed. I will look into what's going on there...

anthony-kolesov commented:
Seems like regression pass without problems. I've pushed patches to a new arc-2017.09 branch. I will submit upstream later.
